editor/plugins/visual_shader_editor_plugin.cpp

```diff
@@ -1835,7 +1835,15 @@ void VisualShaderEditor::_remove_input_port(int p_node, int p_port) {
 	}
 
 	undo_redo->add_do_method(node.ptr(), "remove_input_port", p_port);
-	undo_redo->add_undo_method(node.ptr(), "add_input_port", p_port, (int)node->get_input_port_type(p_port), node->get_input_port_name(p_port));
+	if (p_port < node->get_input_port_count() - 1) {
+		for (int i = p_port; i < node->get_input_port_count() - 1; i++) {
+			undo_redo->add_undo_method(node.ptr(), "set_input_port_name", i, node->get_input_port_name(i));
+			undo_redo->add_undo_method(node.ptr(), "set_input_port_type", i, node->get_input_port_type(i));
+		}
+		undo_redo->add_undo_method(node.ptr(), "add_input_port", node->get_input_port_count(), (int)node->get_input_port_type(node->get_input_port_count() - 1), node->get_input_port_name(node->get_input_port_count() - 1));
+	} else {
+		undo_redo->add_undo_method(node.ptr(), "add_input_port", p_port, (int)node->get_input_port_type(p_port), node->get_input_port_name(p_port));
+	}
```

Comment:
Just a suggestion, but shouldn't this kind of logic be part of `add_input_port` itself? So that it automatically re-assigns port indices when inserting a port in the middle?

Reply:
Or maybe I need to try?

Reply:
@akien-mga Done, I think - I'm surprised it works just by removing a few checks.